File: connector/src/types/feature-request.ts
export type FeatureRequestType = "feature" | "bug_fix" | "improvement" | "integration" | "ux_change";
export type FeatureRequestStatus =
  | "draft"
  | "reviewed"
  | "approved"
  | "rejected"
  | "merged"
  | "sent_to_agent";
export type FeatureRequestPriority = "low" | "medium" | "high" | "critical";

export interface SupportingEvidence {
  signal_id: string;
  signal_summary: string;
  source: string;
  source_data_type: string;
  customer_company?: string | null;
  author_name?: string | null;
  representative_quote: string;
  relevance_score: number;
}

export interface FeatureRequestImage {
  r2_key: string;
  signal_id: string;
  description: string;
  mime_type: string;
}

export interface ImpactMetrics {
  signal_count: number;
  unique_customers: number;
  unique_companies: number;
  source_breakdown: Record<string, number>;
  avg_urgency_score: number;
  avg_sentiment_score: number;
  earliest_mention: string;
  latest_mention: string;
  trend_direction: "increasing" | "stable" | "decreasing" | string;
}

export interface FeatureRequest {
  id: string;
  organization_id: string;
  title: string;
  type: FeatureRequestType;
  status: FeatureRequestStatus;
  priority: FeatureRequestPriority;
  priority_score: number;
  problem_statement: string;
  proposed_solution: string;
  user_story: string;
  acceptance_criteria: string[];
  technical_notes?: string | null;
  affected_product_areas: string[];
  supporting_evidence: SupportingEvidence[];
  images: FeatureRequestImage[];
  impact_metrics?: ImpactMetrics | null;
  synthesis_run_id?: string | null;
  synthesis_model?: string | null;
  synthesis_confidence?: number | null;
  merged_into_id?: string | null;
  human_edited: boolean;
  human_edited_fields: string[];
  human_notes?: string | null;
  created_at: string;
  updated_at: string;
}

export interface FeatureRequestFilters {
  page?: number;
  limit?: number;
  status?: FeatureRequestStatus;
  type?: FeatureRequestType;
  priority?: FeatureRequestPriority;
  min_score?: number;
  signal_id?: string;
  sort?: string;
  order?: "asc" | "desc";
  synthesis_run_id?: string;
}

export interface FeatureRequestPatch {
  title?: string;
  type?: FeatureRequestType;
  priority?: FeatureRequestPriority;
  problem_statement?: string;
  proposed_solution?: string;
  user_story?: string;
  acceptance_criteria?: string[];
  technical_notes?: string;
  affected_product_areas?: string[];
  human_notes?: string;
}

File: synthesis/src/schemas/feature_request.py
from datetime import datetime
from enum import StrEnum
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class FeatureRequestType(StrEnum):
    FEATURE = "feature"
    BUG_FIX = "bug_fix"
    IMPROVEMENT = "improvement"
    INTEGRATION = "integration"
    UX_CHANGE = "ux_change"


class FeatureRequestStatus(StrEnum):
    DRAFT = "draft"
    REVIEWED = "reviewed"
    APPROVED = "approved"
    REJECTED = "rejected"
    MERGED = "merged"
    SENT_TO_AGENT = "sent_to_agent"


class FeatureRequestPriority(StrEnum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class SupportingEvidence(BaseModel):
    signal_id: UUID
    signal_summary: str
    source: str
    source_data_type: str
    customer_company: str | None = None
    author_name: str | None = None
    representative_quote: str
    relevance_score: float = Field(ge=0.0, le=1.0)


class FeatureRequestImage(BaseModel):
    r2_key: str
    signal_id: UUID
    description: str
    mime_type: str


class ImpactMetrics(BaseModel):
    signal_count: int
    unique_customers: int
    unique_companies: int
    source_breakdown: dict[str, int]
    avg_urgency_score: float
    avg_sentiment_score: float
    earliest_mention: datetime
    latest_mention: datetime
    trend_direction: str


class FeatureRequestRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    title: str
    type: FeatureRequestType
    status: FeatureRequestStatus
    priority: FeatureRequestPriority
    priority_score: int
    problem_statement: str
    proposed_solution: str
    user_story: str
    acceptance_criteria: list[str] = Field(default_factory=list)
    technical_notes: str | None = None
    affected_product_areas: list[str] = Field(default_factory=list)
    supporting_evidence: list[SupportingEvidence] = Field(default_factory=list)
    images: list[FeatureRequestImage] = Field(default_factory=list)
    impact_metrics: ImpactMetrics | None = None
    synthesis_run_id: UUID | None = None
    synthesis_model: str | None = None
    synthesis_confidence: int | None = None
    synthesis_summary: str | None = None
    merged_into_id: UUID | None = None
    human_edited: bool
    human_edited_fields: list[str] = Field(default_factory=list)
    human_notes: str | None = None
    created_at: datetime
    updated_at: datetime


class FeatureRequestPatch(BaseModel):
    title: str | None = None
    type: FeatureRequestType | None = None
    priority: FeatureRequestPriority | None = None
    problem_statement: str | None = None
    proposed_solution: str | None = None
    user_story: str | None = None
    acceptance_criteria: list[str] | None = None
    technical_notes: str | None = None
    affected_product_areas: list[str] | None = None
    human_notes: str | None = None


class MergeFeatureRequestRequest(BaseModel):
    target_id: UUID


class FeatureRequestImageUrl(BaseModel):
    r2_key: str
    url: str
    signal_id: UUID
    description: str
    mime_type: str
